Mage.Sets/src/mage/cards/b/BalefulMastery.java

```diff
@@ -0,0 +1,86 @@
+package mage.cards.b;
+
+import mage.abilities.Ability;
+import mage.abilities.costs.AlternativeCostSourceAbility;
+import mage.abilities.costs.mana.ManaCostsImpl;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.ExileTargetEffect;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.Outcome;
+import mage.game.Game;
+import mage.game.stack.Spell;
+import mage.players.Player;
+import mage.target.common.TargetCreatureOrPlaneswalker;
+import mage.target.common.TargetOpponent;
+
+import java.util.UUID;
+
+/**
+ *
+ * @author htrajan
+ */
+public final class BalefulMastery extends CardImpl {
+
+    public BalefulMastery(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.INSTANT}, "{3}{B}");
+
+        // You may pay {1}{B} rather than pay this spell's mana cost.
+        this.addAbility(new AlternativeCostSourceAbility(new ManaCostsImpl<>("{1}{B}")));
+
+        // If the {1}{B} cost was paid, an opponent draws a card.
+        this.getSpellAbility().addEffect(new BalefulMasteryAlternativeCostEffect());
+
+        // Exile target creature or planeswalker.
+        this.getSpellAbility().addEffect(new ExileTargetEffect());
+        this.getSpellAbility().addTarget(new TargetCreatureOrPlaneswalker());
+    }
+
+    private BalefulMastery(final BalefulMastery card) {
+        super(card);
+    }
+
+    @Override
+    public BalefulMastery copy() {
+        return new BalefulMastery(this);
+    }
+}
+
+class BalefulMasteryAlternativeCostEffect extends OneShotEffect {
+
+    BalefulMasteryAlternativeCostEffect() {
+        super(Outcome.Detriment);
+        staticText = "if the {1}{B} cost was paid, an opponent draws a card.<br>";
+    }
+
+    BalefulMasteryAlternativeCostEffect(BalefulMasteryAlternativeCostEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public BalefulMasteryAlternativeCostEffect copy() {
+        return new BalefulMasteryAlternativeCostEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Spell spell = game.getSpell(source.getOriginalId());
+        AlternativeCostSourceAbility altCostAbility = (AlternativeCostSourceAbility) spell.getAbilities().stream()
+            .filter(ability -> ability instanceof AlternativeCostSourceAbility)
+            .findFirst()
+            .orElseThrow(IllegalStateException::new);
+        Object value = game.getState().getValue(altCostAbility.getId().toString());
+        if (value != null && value.toString().equals("ALT_COST_PAID")) {
```

Comment:
Use standard values notation: "activation_key + source.getSourceId()".

Reply:
I don't like your workaround with whole `AlternativeCostSourceAbility` (ALT_COST_PAID) -- it's potentially bugged. Card can have multiple alternative cost, even dynamic added. Any can be activated, but you must search only specific cost from the card. That's why `ability.originalId` exists.

See example:
![shot_210406_214158](https://user-images.githubusercontent.com/8344157/113755313-3b53e180-9721-11eb-9352-76d7c6aff11b.png)

Reply:
Ah ok got it -- I added an INFO_KEY to `AlternativeCostSourceAbility` and am storing the value now using `ability.originalId + INFO_KEY`. Not sure if this is what you had in mind.

Reply:
Nope, look at my example -- you must save `originalId` on `BalefulMasteryAlternativeCostEffect` creation. So that effect can find real alternative cost on resolve. Maybe ZCC check needs too (zone change counter -- if card returns to hand after first usage). Look at `MadnessCastEffect` and `MadnessCondition` implementations for real spell searching.

Code like that:
1. Init card with abilities and save AlternativeCostSourceAbility's `originalId` in the `BalefulMasteryAlternativeCostEffect` effect;
2. Check cost conditional like madness example (find spell by `getLastKnownInformation` with zcc and check cost status) for effect.

Reply:
I'll fix/improve it myself.